mlflow/pyfunc/__init__.py

```diff
@@ -799,9 +800,22 @@ def _predict(self, data: PyFuncInput, params: Optional[dict[str, Any]] = None) -
         # fetch the schema from metadata to avoid signature change after model is loaded
         self.input_schema = self.metadata.get_input_schema()
         self.params_schema = self.metadata.get_params_schema()
-        data, params = _validate_prediction_input(
-            data, params, self.input_schema, self.params_schema, self.loader_module
-        )
+        # signature can only be inferred from type hints if the model is PythonModel
+        if self.metadata._is_signature_from_type_hint():
+            from mlflow.types.type_hints import (
+                _convert_data_to_type_hint,
+                _validate_example_against_type_hint,
+            )
+
+            type_hints = self._model_impl.python_model._get_type_hints()
+            data = _convert_data_to_type_hint(data, type_hints.input)
+            # TODO: remove the validation here once we move the logic inside PythonModel
+            data = _validate_example_against_type_hint(data, type_hints.input)
```

Comment:
nit: unrelated to this PR, but we may want to reconsider the naming of this method. It is not natural to expect the "validata_..." method to convert the data and "data" here is not an example in this flow.

Reply:
Yea... but we only convert the data for pydantic model type hint, so _convert_data_to_type_hint sounds a bit weird as well 🤔